posts/change-data-capture-cloudsql-postgres-with-apache-airflow-1.md

```diff
@@ -0,0 +1,1016 @@
+authors:
+  - Rob Deeb
+date: 2020-01-21T12:00:00.000Z
+---
+> **Note:** All of the code in this post is available in [this Github repository](https://github.com/astronomer/cdc-cloudsql-airflow-demo) and can be run locally using the Astronomer CLI.
+
+Extracting changes in data from a database is often needed to keep that data current in separate environments. It's especially important in contexts such as populating a search engine, training machine learning models, or building analytic reports from a warehouse. This post walks you through a scheduled method for extracting data changes using Apache Airflow in a way that is near-real time, flexible, and highly available.
+
+## Overview
+
+The scheduled approach outlined in this post - enabling change data to be updated in near-real time - has the added benefit of built-in data partitioning formats directly from the process. In a distributed data network, using Airflow as the scheduled approach pushes it into the realm of Consistency and High Availability (CA) in the CAP theorem. That is, you can base SLAs on this approach.
+
+Those capturing data in the format of time-series or events will benefit the most with this approach, but it can be used for any data by adding soft deletes to your transactions and defining the time scope to which changes must be recorded. Eventual consistency is achieved in a robust manner by creating an idempotent approach to how the data is extracted and moved. 
+
+In this series, we'll outline all steps in the process from end to end, from setup to extract. In totality, this will include topics of:
+
+- Setting up security and IAM in GCP
+- Calling the CloudSQL Export API
+- Setting up Airflow and connecting it to GCP
+- Building a DAG in Airflow
+- Configuring Watermarking for data extract intervals
+- Writing an operator in Airflow to extract the data
+- Integrating the operator into the DAG
+- Scheduling and coordinating the extract
+- Deployment to Production on Astronomer Cloud
+
+In a later post, I will show you how to transform or clean the data then add it to a warehouse as part of the same workflow. We will be using the [Astronomer Platform](https://www.astronomer.io/), as it is the easiest way to get your Apache Airflow instance up and running.
+
+## How it Works
+
+The base method for extract uses the [CloudSQL Export API](https://cloud.google.com/sql/docs/postgres/import-export/exporting#rest-v1beta4). This runs an operation with a SQL query, sending the results to a Google Cloud Storage (GCS) bucket as declared by you. You could just as easily extract directly from your database if your system and workload allows for such. 
+
+Using the extract API provides a robust and easily configurable security layer controlled by a service account, such that no system outside of CloudSQL is touching your production database. It also provides a "serverless offload" option in the case that you need to guarantee zero load on your database from the extract. 
```

Comment:
```suggestion
Using the extract API provides a robust and easily configurable security layer controlled by a service account. This ensures that no system outside of CloudSQL is touching your production database. It also provides a "serverless offload" option in case you need to guarantee zero load on your database from the extract. 
```

Reply:
@jwitz "In case" feels different. Is there another way we can say it? Saying "in case you need" expresses that option as more of a backup plan instead of a specific use-case. Some people may have such a big extract to do, they won't want to push on their DB, and instead opt to use the serverless offload.

Reply:
Let's keep it simple then:
```suggestion
Using the extract API provides a robust and easily configurable security layer controlled by a service account, such that no system outside of CloudSQL is touching your production database. It also provides a "serverless offload" option if you need to guarantee zero load on your database from the extract. 
```

Reply:
How about "It also provides a "serverless offload" option in case you need to guarantee zero load on your database from the extract"